subprojects/core/src/main/java/org/gradle/api/services/internal/DefaultBuildServicesRegistry.java

```
@@ -118,40 +122,82 @@ public SharedResource forService(BuildServiceProvider<?, ?> service) {
+        return doRegisterIfAbsent(name, implementationType, () -> {
             // TODO - extract some shared infrastructure to take care of parameter instantiation (eg strict vs lenient, which services are visible)
             P parameters = instantiateParametersOf(implementationType);
 
             // TODO - should defer execution of the action, to match behaviour for other container `register()` methods.
-
             DefaultServiceSpec<P> spec = uncheckedNonnullCast(specInstantiator.newInstance(DefaultServiceSpec.class, parameters));
             configureAction.execute(spec);
-            Integer maxParallelUsages = spec.getMaxParallelUsages().getOrNull();
+            return spec;
+        });
+    }
+
+    @Override
+    public BuildServiceProvider<?, ?> registerIfAbsent(String name, Class<? extends BuildService<?>> implementationType, @Nullable BuildServiceParameters parameters, int maxUsages) {
+        Supplier<BuildServiceSpec<?>> buildServiceSpecSupplier = () -> {
+            DefaultServiceSpec<?> spec = uncheckedNonnullCast(specInstantiator.newInstance(DefaultServiceSpec.class, parameters));
+            spec.getMaxParallelUsages().set(maxUsages);
+            return spec;
+        };
+        return doRegisterIfAbsent(name, uncheckedNonnullCast(implementationType), uncheckedNonnullCast(buildServiceSpecSupplier));
+    }
 
+    private <T extends BuildService<P>, P extends BuildServiceParameters> BuildServiceProvider<T, P> doRegisterIfAbsent(String name, Class<T> implementationType, Supplier<BuildServiceSpec<P>> specSupplier) {
+        return withRegistrations(registrations -> {
+            BuildServiceRegistration<?, ?> existing = registrations.findByName(name);
+            if (existing != null) {
+                Class existingType = ((ProviderInternal) existing.getService()).getType();
+                if (!implementationType.isAssignableFrom(existingType)) {
```

As discussed during the daily call, we now fail early if you try to re-register a service using a different type (instead of failing with a ClassCastException, which may be cryptic if we are talking same class names but different classes)